pkg/kube/krt/collection.go

```diff
@@ -28,28 +28,54 @@ import (
 	"istio.io/istio/pkg/util/sets"
 )
 
+type indexedDependencyType uint8
+
+const (
+	unknownIndexType indexedDependencyType = iota
+	indexType        indexedDependencyType = iota
+	getKeyType       indexedDependencyType = iota
+)
+
+var allIndexedDependencyTypes = []indexedDependencyType{indexType, getKeyType}
+
 type dependencyState[I any] struct {
 	// collectionDependencies specifies the set of collections we depend on from within the transformation functions (via Fetch).
 	// These are keyed by the internal uid() function on collections.
 	// Note this does not include `parent`, which is the *primary* dependency declared outside of transformation functions.
 	collectionDependencies sets.Set[collectionUID]
 	// Stores a map of I -> secondary dependencies (added via Fetch)
-	objectDependencies           map[Key[I]][]*dependency
-	indexedDependencies          map[indexedDependency]sets.Set[Key[I]]
-	indexedDependenciesExtractor map[collectionUID]objectKeyExtractor
+	objectDependencies  map[Key[I]][]*dependency
+	indexedDependencies map[indexedDependency]sets.Set[Key[I]]
+	// indexedDependenciesExtractor stores a map of [collection,fetch type] => an extractor to get change keys.
+	// Note that a given collection can have multiple Fetches, but they are limited to those of the different kind.
+	// I.e. you can do a `Fetch(c1, FilterIndex()) + Fetch(c1, FilterKey())` but not `Fetch(c1, FilterIndex()) + Fetch(c1, FilterIndex())`.
```

Comment:
Comment seems scarier than I think it really is. This limitation on duplicate fetches on the same secondary with the same indexedDependencyType is only within a single transformation. If C2 and C3 each include an filter indexed fetch of C1 that should be fine though.

Reply:
Yeah that is accurate

Reply:
I updated the comment

Reply:
Nice, I want to say that was the only little nit I had. I'll take another look shortly